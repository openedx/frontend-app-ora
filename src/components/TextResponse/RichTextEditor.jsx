import React from 'react';
import PropTypes from 'prop-types';

import { Editor } from '@tinymce/tinymce-react';
import 'tinymce/tinymce.min';
import 'tinymce/icons/default';
import 'tinymce/plugins/link';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/code';
import 'tinymce/plugins/image';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.min.css';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import { useIntl } from '@edx/frontend-platform/i18n';
import messages from './messages';

export const stateKeys = StrictDict({
  value: 'value',
});

const RichTextEditor = ({
  // id,
  initialValue,
  disabled,
  optional,
}) => {
  const [value, setValue] = useKeyedState(stateKeys.value, initialValue);
  const { formatMessage } = useIntl();

  const extraConfig = disabled ? {
    toolbar: false,
    readonly: 1,
  } : {
    // eslint-disable-next-line max-len
    toolbar: 'formatselect | bold italic underline | link blockquote image | numlist bullist outdent indent | strikethrough | code | undo redo',
  };

  return (
    <div className="form-group">
      <label htmlFor="rich-text-response">
        {formatMessage(messages.yourResponse)} ({formatMessage(optional ? messages.optional : messages.required)})
      </label>
      <Editor
        name="rich-text-response"
        initialValue={value}
        init={{
          menubar: false,
          statusbar: false,
          skin: false,
          content_css: false,
          height: '300',
          schema: 'html5',
          plugins: 'code image link lists',
          ...extraConfig,
        }}
        onChange={(e) => setValue(e.target.getContent())}
        disabled={disabled}
      />
    </div>
  );
};

RichTextEditor.defaultProps = {
  disabled: false,
  initialValue: '',
  optional: false,
};

RichTextEditor.propTypes = {
  // id: PropTypes.string.isRequired,
  input: PropTypes.shape({
    value: PropTypes.string,
    name: PropTypes.string,
    onChange: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    touched: PropTypes.bool,
    submitFailed: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
  disabled: PropTypes.bool,
  initialValue: PropTypes.string,
  optional: PropTypes.bool,
};

export default RichTextEditor;
