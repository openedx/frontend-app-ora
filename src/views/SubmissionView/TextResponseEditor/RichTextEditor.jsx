import React from 'react';
import PropTypes from 'prop-types';

import { Editor } from '@tinymce/tinymce-react';
import 'tinymce/tinymce.min';
import 'tinymce/icons/default';
import 'tinymce/plugins/link';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/code';
import 'tinymce/plugins/image';
import 'tinymce/themes/silver';

import { useIntl } from '@edx/frontend-platform/i18n';
import messages from './messages';

const RichTextEditor = ({
  // id,
  value,
  disabled,
  optional,
  onChange,
}) => {
  const { formatMessage } = useIntl();

  const extraConfig = disabled ? {
    toolbar: false,
    readonly: 1,
  } : {
    // eslint-disable-next-line max-len
    toolbar: 'formatselect | bold italic underline | link blockquote image | numlist bullist outdent indent | strikethrough | code | undo redo',
  };

  return (
    <div className="form-group ora-tinymce">
      <label htmlFor="rich-text-response h3">
        {formatMessage(messages.yourResponse)} ({formatMessage(optional ? messages.optional : messages.required)})
      </label>
      <Editor
        className="bg-white"
        name="rich-text-response"
        value={value}
        init={{
          menubar: false,
          statusbar: false,
          // TODO: rewrite this to use skin=false and content_css=false when we figure
          //       which part of css-loader to change to stop tinymce from changing the
          //       oxide skin css.
          base_url: `${process.env.LMS_BASE_URL}/static/js/vendor/tinymce/js/tinymce`,
          skin: 'studio-tmce5',
          content_css: 'studio-tmce5',
          height: '300',
          schema: 'html5',
          plugins: 'code image link lists',
          ...extraConfig,
        }}
        onEditorChange={onChange}
        disabled={disabled}
      />
    </div>
  );
};

RichTextEditor.defaultProps = {
  disabled: false,
  value: '',
  optional: false,
  onChange: () => { },
};

RichTextEditor.propTypes = {
  // id: PropTypes.string.isRequired,
  // input: PropTypes.shape({
  //   value: PropTypes.string,
  //   name: PropTypes.string,
  //   onChange: PropTypes.func.isRequired,
  // }).isRequired,
  // meta: PropTypes.shape({
  //   touched: PropTypes.bool,
  //   submitFailed: PropTypes.bool,
  //   error: PropTypes.string,
  // }).isRequired,
  disabled: PropTypes.bool,
  value: PropTypes.string,
  optional: PropTypes.bool,
  onChange: PropTypes.func,
};

export default RichTextEditor;
