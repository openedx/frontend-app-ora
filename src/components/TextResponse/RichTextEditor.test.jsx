import { shallow } from '@edx/react-unit-test-utils';
import RichTextEditor from './RichTextEditor';

jest.mock('@tinymce/tinymce-react', () => ({
  Editor: () => 'Editor',
}));

jest.mock('tinymce/tinymce.min', () => 'tinymce');
jest.mock('tinymce/icons/default', () => 'default');
jest.mock('tinymce/plugins/link', () => 'link');
jest.mock('tinymce/plugins/lists', () => 'lists');
jest.mock('tinymce/plugins/code', () => 'code');
jest.mock('tinymce/plugins/image', () => 'image');
jest.mock('tinymce/themes/silver', () => 'silver');

describe('<RichTextEditor />', () => {
  const props = {
    optional: true,
    disabled: false,
    value: 'value',
    onChange: jest.fn().mockName('onChange'),
  };

  it('render optional', () => {
    const wrapper = shallow(<RichTextEditor {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('label')[0].el.children).toContain('Optional');
    expect(wrapper.instance.findByType('Editor')[0].props.init.readonly).not.toEqual(1);
  });

  it('render required', () => {
    const wrapper = shallow(<RichTextEditor {...props} optional={false} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('label')[0].el.children).toContain('Required');
  });

  it('render disabled', () => {
    const wrapper = shallow(<RichTextEditor {...props} disabled />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Editor')[0].props.init.readonly).toEqual(1);
  });
});
