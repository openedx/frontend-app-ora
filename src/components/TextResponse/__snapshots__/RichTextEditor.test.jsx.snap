// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichTextEditor /> render disabled 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="rich-text-response"
  >
    Your response
     (
    Optional
    )
  </label>
  <Editor
    disabled={true}
    init={
      Object {
        "content_css": false,
        "height": "300",
        "menubar": false,
        "plugins": "code image link lists",
        "readonly": 1,
        "schema": "html5",
        "skin": false,
        "statusbar": false,
        "toolbar": false,
      }
    }
    name="rich-text-response"
    onEditorChange={[MockFunction onChange]}
    value="value"
  />
</div>
`;

exports[`<RichTextEditor /> render optional 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="rich-text-response"
  >
    Your response
     (
    Optional
    )
  </label>
  <Editor
    disabled={false}
    init={
      Object {
        "content_css": false,
        "height": "300",
        "menubar": false,
        "plugins": "code image link lists",
        "schema": "html5",
        "skin": false,
        "statusbar": false,
        "toolbar": "formatselect | bold italic underline | link blockquote image | numlist bullist outdent indent | strikethrough | code | undo redo",
      }
    }
    name="rich-text-response"
    onEditorChange={[MockFunction onChange]}
    value="value"
  />
</div>
`;

exports[`<RichTextEditor /> render required 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="rich-text-response"
  >
    Your response
     (
    Required
    )
  </label>
  <Editor
    disabled={false}
    init={
      Object {
        "content_css": false,
        "height": "300",
        "menubar": false,
        "plugins": "code image link lists",
        "schema": "html5",
        "skin": false,
        "statusbar": false,
        "toolbar": "formatselect | bold italic underline | link blockquote image | numlist bullist outdent indent | strikethrough | code | undo redo",
      }
    }
    name="rich-text-response"
    onEditorChange={[MockFunction onChange]}
    value="value"
  />
</div>
`;
